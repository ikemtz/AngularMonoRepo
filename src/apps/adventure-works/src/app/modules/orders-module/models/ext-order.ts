import { PagerSettings } from '@progress/kendo-angular-grid';
import { ODataResult, ODataState } from 'imng-kendo-odata';
import { IOrder, IOrderLineItem } from '../../../models/odata';

export interface IExtOrder extends IOrder {
  orderLineItemODataState: ODataState;
  orderLineItemOData: ODataResult<IOrderLineItem>;
  orderLineItemPagerSettings: PagerSettings | false;
}
