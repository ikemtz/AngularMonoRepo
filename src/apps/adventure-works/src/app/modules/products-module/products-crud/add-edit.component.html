<imng-data-entry-dialog
  [width]="700"
  [height]="550"
  [parentComponent]="this"
  autoFocusedElement="#name">
  <form novalidate [formGroup]="addEditForm" autocomplete="false" class="row">
    <div class="col-md-6">
      <label for="id" class="control-label mr-4">Id:</label>
      <input
        id="id"
        type="text"
        class="form-control"
        [formControlName]="props.ID"
        [ngClass]="{
          'is-invalid': (submitted$ | async) && formControlErrors(props.ID)
        }" />
      <div
        *ngIf="(submitted$ | async) && formControlErrors(props.ID)"
        class="invalid-feedback">
        <div *ngIf="formControlErrors(props.ID)?.[RequiredError]">
          Product id is required
        </div>
      </div>
    </div>
    <div class="col-md-6">
      <label for="name" class="control-label mr-4">Name:</label>
      <input
        id="name"
        type="text"
        maxlength="50"
        class="form-control"
        [formControlName]="props.NAME"
        [ngClass]="{
          'is-invalid': (submitted$ | async) && formControlErrors(props.NAME)
        }" />
      <div
        *ngIf="(submitted$ | async) && formControlErrors(props.NAME)"
        class="invalid-feedback">
        <div *ngIf="formControlErrors(props.NAME)?.[RequiredError]">
          Product name is required
        </div>
      </div>
    </div>
    <div class="col-md-6">
      <label for="num" class="control-label mr-4">Num:</label>
      <input
        id="num"
        type="text"
        maxlength="25"
        class="form-control"
        [formControlName]="props.NUM"
        [ngClass]="{
          'is-invalid': (submitted$ | async) && formControlErrors(props.NUM)
        }" />
      <div
        *ngIf="(submitted$ | async) && formControlErrors(props.NUM)"
        class="invalid-feedback">
        <div *ngIf="formControlErrors(props.NUM)?.[RequiredError]">
          Product num is required
        </div>
      </div>
    </div>
    <div class="col-md-6">
      <label for="color" class="control-label mr-4">Color:</label>
      <input
        id="color"
        type="text"
        maxlength="15"
        class="form-control"
        [formControlName]="props.COLOR" />
    </div>
    <div class="col-md-6">
      <label for="standard_cost" class="control-label mr-4"
        >Standard Cost:</label
      >
      <input
        id="standard_cost"
        type="number"
        class="form-control"
        [formControlName]="props.STANDARD_COST"
        [ngClass]="{
          'is-invalid':
            (submitted$ | async) && formControlErrors(props.STANDARD_COST)
        }" />
      <div
        *ngIf="(submitted$ | async) && formControlErrors(props.STANDARD_COST)"
        class="invalid-feedback">
        <div *ngIf="formControlErrors(props.STANDARD_COST)?.[RequiredError]">
          Product standard cost is required
        </div>
      </div>
    </div>
    <div class="col-md-6">
      <label for="list_price" class="control-label mr-4">List Price:</label>
      <input
        id="list_price"
        type="number"
        class="form-control"
        [formControlName]="props.LIST_PRICE"
        [ngClass]="{
          'is-invalid':
            (submitted$ | async) && formControlErrors(props.LIST_PRICE)
        }" />
      <div
        *ngIf="(submitted$ | async) && formControlErrors(props.LIST_PRICE)"
        class="invalid-feedback">
        <div *ngIf="formControlErrors(props.LIST_PRICE)?.[RequiredError]">
          Product list price is required
        </div>
      </div>
    </div>
    <div class="col-md-6">
      <label for="size" class="control-label mr-4">Size:</label>
      <input
        id="size"
        type="text"
        maxlength="5"
        class="form-control"
        [formControlName]="props.SIZE" />
    </div>
    <div class="col-md-6">
      <label for="weight" class="control-label mr-4">Weight:</label>
      <input
        id="weight"
        type="number"
        class="form-control"
        [formControlName]="props.WEIGHT" />
    </div>
    <div class="col-md-6">
      <label for="product_category_id" class="control-label mr-4"
        >Product Category Id:</label
      >
      <input
        id="product_category_id"
        type="text"
        class="form-control"
        [formControlName]="props.PRODUCT_CATEGORY_ID" />
    </div>
    <div class="col-md-6">
      <label for="product_model_id" class="control-label mr-4"
        >Product Model Id:</label
      >
      <input
        id="product_model_id"
        type="text"
        class="form-control"
        [formControlName]="props.PRODUCT_MODEL_ID" />
    </div>
    <div class="col-md-6">
      <label for="sell_start_date" class="control-label mr-4"
        >Sell Start Date:</label
      >
      <kendo-datepicker
        id="sell_start_date"
        class="form-control"
        [formControlName]="props.SELL_START_DATE"
        [ngClass]="{
          'is-invalid':
            (submitted$ | async) && formControlErrors(props.SELL_START_DATE)
        }"></kendo-datepicker>
      <div
        *ngIf="(submitted$ | async) && formControlErrors(props.SELL_START_DATE)"
        class="invalid-feedback">
        <div *ngIf="formControlErrors(props.SELL_START_DATE)?.[RequiredError]">
          Product sell start date is required
        </div>
      </div>
    </div>
    <div class="col-md-6">
      <label for="sell_end_date" class="control-label mr-4"
        >Sell End Date:</label
      >
      <kendo-datepicker
        id="sell_end_date"
        class="form-control"
        [formControlName]="props.SELL_END_DATE"></kendo-datepicker>
    </div>
    <div class="col-md-6">
      <label for="discontinued_date" class="control-label mr-4"
        >Discontinued Date:</label
      >
      <kendo-datepicker
        id="discontinued_date"
        class="form-control"
        [formControlName]="props.DISCONTINUED_DATE"></kendo-datepicker>
    </div>
    <div class="col-md-6">
      <label for="thumb_nail_photo" class="control-label mr-4"
        >Thumb Nail Photo:</label
      >
      <input
        id="thumb_nail_photo"
        type="text"
        maxlength="5000"
        class="form-control"
        [formControlName]="props.THUMB_NAIL_PHOTO" />
    </div>
    <div class="col-md-6">
      <label for="product_model" class="control-label mr-4"
        >Product Model:</label
      >
      <kendo-multicolumncombobox
        id="product_model"
        placeholder="Select a product model..."
        [formControlName]="props.PRODUCT_MODEL_ID"
        [data]="productModels$ | async"
        [textField]="productModelProps.NAME"
        [valueField]="productModelProps.ID"
        [valuePrimitive]="true"
        [filterable]="true"
        (filterChange)="handleProductModelFilter($event)">
        <kendo-combobox-column
          field="name"
          title="Name"
          [width]="300"></kendo-combobox-column>
        <kendo-combobox-column
          field="description"
          title="Description"
          [width]="300"></kendo-combobox-column>
      </kendo-multicolumncombobox>
      <div *ngIf="formControlErrors(props.PRODUCT_MODEL_ID)?.[RequiredError]">
        Product model is required
      </div>
    </div>
    <div class="col-md-6">
      <label for="product_category" class="control-label mr-4"
        >Product Category:</label
      >
      <kendo-combobox
        id="product_category"
        placeholder="Select a product category..."
        [formControlName]="props.PRODUCT_CATEGORY_ID"
        [data]="productCategories$ | async"
        [textField]="productCategoryProps.NAME"
        [valueField]="productCategoryProps.ID"
        [valuePrimitive]="true"
        [filterable]="true"
        (filterChange)="handleProductCategoryFilter($event)">
      </kendo-combobox>
      <div
        *ngIf="formControlErrors(props.PRODUCT_CATEGORY_ID)?.[RequiredError]">
        Product category is required
      </div>
    </div>
  </form>
</imng-data-entry-dialog>
