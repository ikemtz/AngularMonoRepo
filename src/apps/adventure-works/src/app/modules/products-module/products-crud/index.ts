export { ProductAddComponent } from './add.component';
export { ProductEditComponent } from './edit.component';
export { ProductApiService } from './api.service';
export { ProductCrudFacade } from './crud.facade';
