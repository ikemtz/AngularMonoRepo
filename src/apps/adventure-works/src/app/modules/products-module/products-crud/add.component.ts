import {
  Component,
  ChangeDetectionStrategy,
  OnInit,
  OnDestroy,
} from '@angular/core';

import { ProductCrudFacade } from './crud.facade';
import { ProductBaseEntryComponent } from './base-entry.component';
import { IProduct } from '../../../models/webapi';
import { normalizeRequest } from 'imng-nrsrx-client-utils';

@Component({
  selector: 'aw-product-add',
  templateUrl: './add-edit.component.html',
  styleUrls: ['./add-edit.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ProductAddComponent
  extends ProductBaseEntryComponent
  implements OnInit, OnDestroy {
  public dialogTitle = 'Add Product';
  public active$ = this.facade.isNewActive$;

  constructor(facade: ProductCrudFacade) {
    super(facade);
  }
  public override initForm(): void {
    super.initForm();
    this.addEditForm.patchValue({});
  }

  public save(): void {
    const val = normalizeRequest<IProduct>(this.addEditForm.value);
    val.id = undefined;
    this.facade.saveNewEntity(val);
  }
}
