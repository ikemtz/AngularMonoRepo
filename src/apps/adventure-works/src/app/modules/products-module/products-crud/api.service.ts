import { Injectable } from '@angular/core';
import { NrsrxBaseApiClientService } from 'imng-nrsrx-client-utils';
import { HttpClient } from '@angular/common/http';
import { IProduct } from '../../../models';
import { environment } from '../../../../environments/environment';

@Injectable({
  providedIn: 'root',
})
export class ProductApiService extends NrsrxBaseApiClientService<IProduct> {
  public url = environment.endPoints.products;
  constructor(http: HttpClient) {
    super(http);
  }
}
