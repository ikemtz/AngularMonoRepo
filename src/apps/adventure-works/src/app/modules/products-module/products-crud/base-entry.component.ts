import { OnInit, Component } from '@angular/core';
import { FormGroup } from '@angular/forms';
import { BaseDataEntryComponent } from 'imng-kendo-data-entry';
import { BehaviorSubject, map, Observable, switchMap } from 'rxjs';
import { IProductForm } from '../../../models/odata';
import {
  IProductCategory,
  IProductModel,
  ProductCategoryProperties,
  ProductFormGroupFac,
  ProductModelProperties,
  ProductProperties,
} from '../../../models/webapi';

import { ProductCrudFacade } from './crud.facade';

@Component({
    template: '',
    standalone: false
})
export abstract class ProductBaseEntryComponent
  extends BaseDataEntryComponent<ProductCrudFacade>
  implements OnInit
{
  public readonly props = ProductProperties;
  public readonly productModelProps = ProductModelProperties;
  public readonly productModels$: Observable<IProductModel[]>;
  public readonly productModelFilter$ = new BehaviorSubject('');
  public readonly productCategoryProps = ProductCategoryProperties;
  public readonly productCategories$: Observable<IProductCategory[]>;
  public readonly productCategoryFilter$ = new BehaviorSubject('');
  public addEditForm: FormGroup<IProductForm>;

  constructor(facade: ProductCrudFacade) {
    super(facade);
    this.productModels$ = facade.productModels$.pipe(
      switchMap((productModels) =>
        this.productModelFilter$.pipe(
          map((productModelFilter) =>
            productModelFilter
              ? productModels.filter(
                  (productModel) =>
                    (productModel.name &&
                      productModel.name
                        .toLowerCase()
                        .indexOf(productModelFilter) >= 0) ||
                    (productModel.description &&
                      productModel.description
                        .toLowerCase()
                        .indexOf(productModelFilter) >= 0),
                )
              : productModels,
          ),
        ),
      ),
    );
    this.productCategories$ = facade.productCategories$.pipe(
      switchMap((productCategories) =>
        this.productCategoryFilter$.pipe(
          map((productCategoryFilter) =>
            productCategoryFilter
              ? productCategories.filter(
                  (productCategory) =>
                    productCategory.name &&
                    productCategory.name
                      .toLowerCase()
                      .indexOf(productCategoryFilter) >= 0,
                )
              : productCategories,
          ),
        ),
      ),
    );
  }

  public override ngOnInit(): void {
    super.ngOnInit();
    this.facade.loadProductModels({
      selectors: [
        ProductModelProperties.ID,
        ProductModelProperties.NAME,
        ProductModelProperties.DESCRIPTION,
      ],
    });
    this.facade.loadProductCategories({
      selectors: [ProductCategoryProperties.ID, ProductCategoryProperties.NAME],
    });
  }

  public initForm(): void {
    this.addEditForm = ProductFormGroupFac();
  }

  public cancel(): void {
    this.facade.clearCurrentEntity();
  }

  public handleProductModelFilter(value: string) {
    this.productModelFilter$.next(value.toLowerCase());
  }
  public handleProductCategoryFilter(value: string) {
    this.productCategoryFilter$.next(value.toLowerCase());
  }
}
