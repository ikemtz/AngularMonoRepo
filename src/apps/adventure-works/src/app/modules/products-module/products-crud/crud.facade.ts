import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';
import { IDataEntryFacade } from 'imng-kendo-data-entry';
import { productsFeature, ProductsPartialState } from '../+state/product.reducer';
import { productQueries } from '../+state/product.selectors';
import * as productActionTypes from '../+state/product.actions';
import { IProduct } from '../../../models/webapi';

@Injectable()
export class ProductCrudFacade implements IDataEntryFacade<IProduct> {
  loading$ = this.store.pipe(select(productsFeature.selectLoading));
  currentEntity$ = this.store.pipe(select(productQueries.getCurrentProduct));
  isEditActive$ = this.store.pipe(select(productQueries.getIsEditProductActive));
  isNewActive$ = this.store.pipe(select(productQueries.getIsNewProductActive));

  constructor(private readonly store: Store<ProductsPartialState>) { }

  public setCurrentEntity(item: IProduct): void {
    this.store.dispatch(productActionTypes.setCurrentProduct(item));
  }

  public clearCurrentEntity(): void {
    this.store.dispatch(productActionTypes.clearCurrentProduct());
  }

  public saveNewEntity(item: IProduct): void {
    this.store.dispatch(productActionTypes.saveProductRequest(item));
  }

  public updateExistingEntity(item: IProduct): void {
    this.store.dispatch(productActionTypes.updateProductRequest(item));
  }
}
