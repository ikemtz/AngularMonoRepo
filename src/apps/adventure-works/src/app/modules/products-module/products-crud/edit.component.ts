import {
  Component,
  ChangeDetectionStrategy,
  OnInit,
  OnDestroy,
} from '@angular/core';
import { formGroupPatcher } from 'imng-kendo-data-entry';
import { normalizeRequest } from 'imng-nrsrx-client-utils';
import { IProduct } from '../../../models/webapi';

import { ProductBaseEntryComponent } from './base-entry.component';
import { ProductCrudFacade } from './crud.facade';

@Component({
  selector: 'aw-product-edit',
  templateUrl: './add-edit.component.html',
  styleUrls: ['./add-edit.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ProductEditComponent
  extends ProductBaseEntryComponent
  implements OnInit, OnDestroy {
  public dialogTitle = 'Edit Product';
  public active$ = this.facade.isEditActive$;

  constructor(facade: ProductCrudFacade) {
    super(facade);
  }
  public override initForm(): void {
    super.initForm();
    if (this.addEditForm) {
      this.allSubscriptions.push(
        this.facade.currentEntity$
          .pipe(formGroupPatcher(this.addEditForm))
          .subscribe(),
      );
    }
  }

  public save(): void {
    const val = normalizeRequest<IProduct>(this.addEditForm.value);
    this.facade.updateExistingEntity(val);
  }
}
