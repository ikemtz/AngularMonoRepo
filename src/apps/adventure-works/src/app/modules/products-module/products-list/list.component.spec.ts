import { CUSTOM_ELEMENTS_SCHEMA, NO_ERRORS_SCHEMA } from '@angular/core';
import { waitForAsync, ComponentFixture, TestBed } from '@angular/core/testing';
import {
  createDataEntryMockFacade,
  createDataDeleteMockFacade,
} from 'imng-kendo-data-entry/testing';
import { createODataGridMockFacade } from 'imng-kendo-grid-odata/testing';

import { ProductListComponent } from './list.component';
import { ProductListFacade } from './list.facade';
import { ProductCrudFacade } from '../products-crud';
import { createTestProduct } from '../../../models/odata';
import { provideRouter } from '@angular/router';
import { productRoutes } from '../products.routing';
import { provideOidcMockFacade } from 'imng-oidc-client/testing';

describe('ProductListComponent', () => {
  let component: ProductListComponent;
  let fixture: ComponentFixture<ProductListComponent>;
  let listFacade: ProductListFacade;
  let crudFacade: ProductCrudFacade;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [ProductListComponent],
      imports: [],
      providers: [
        {
          provide: ProductListFacade,
          useValue: createODataGridMockFacade(createDataDeleteMockFacade()),
        },
        { provide: ProductCrudFacade, useValue: createDataEntryMockFacade() },
        provideRouter(productRoutes),
        provideOidcMockFacade(),
      ],
      schemas: [CUSTOM_ELEMENTS_SCHEMA, NO_ERRORS_SCHEMA],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(ProductListComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    listFacade = TestBed.inject(ProductListFacade);
    crudFacade = TestBed.inject(ProductCrudFacade);
  });

  afterAll(() => {
    component.ngOnDestroy();
  });

  test('it should create', () => {
    expect(component).toBeTruthy();
  });

  test('it should handle DetailExpanded', () => {
    const dataItem = createTestProduct();
    component.detailExpanded({ dataItem } as never);
    expect(component.currentItem).toEqual(dataItem);
  });

  test('it should handle reload', () => {
    component.reloadEntities();
    expect(listFacade.reloadEntities).toHaveBeenCalledTimes(1);
  });

  test('it should handle AddItem', () => {
    component.addItem();
    expect(crudFacade.setCurrentEntity).toHaveBeenCalledTimes(1);
    expect(crudFacade.setCurrentEntity).toHaveBeenCalledWith({});
  });

  test('it should handle EditItem', () => {
    const item = createTestProduct();
    component.editItem(item);
    expect(crudFacade.setCurrentEntity).toHaveBeenCalledTimes(1);
    expect(crudFacade.setCurrentEntity).toHaveBeenCalledWith(item);
  });

  test('it should handle DeleteItem', () => {
    const item = createTestProduct();
    component.deleteItem(item);
    expect(listFacade.deleteExistingEntity).toHaveBeenCalledTimes(1);
    expect(listFacade.deleteExistingEntity).toHaveBeenCalledWith(item);
  });

  test('it should getFormatSrc', () => {
    const result = component.getFormatSrc(
      'iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNkYPhfDwAChwGA60e6kgAAAABJRU5ErkJggg==',
      'png',
    );
    expect(result).toMatchSnapshot();
  });
});
