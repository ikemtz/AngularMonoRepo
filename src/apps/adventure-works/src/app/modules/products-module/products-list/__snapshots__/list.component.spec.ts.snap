// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductListComponent it should getFormatSrc 1`] = `
SafeUrlImpl {
  "changingThisBreaksApplicationSecurity": "data:image/png;base64, iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNkYPhfDwAChwGA60e6kgAAAABJRU5ErkJggg==",
}
`;
